### PR TITLE
[Examples] OCR Ray Datasets example

### PR DESCRIPTION
**Pull request by pcmoritz**
## Why are these changes needed?

This is a simple example that shows how to do OCR with Ray Datasets. It includes:

- How to upload and download the dataset to and from S3
- How to run OCR on the dataset with tesseract
- How to use actors to keep around and re-use a spaCy context for doing NLP on the data

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by pcmoritz** [comment]
@clarkzinzow This is ready now, do you want to review it?

**Comment by pcmoritz** [comment]
Rendered version of the example:  https://ray--25930.org.readthedocs.build/en/25930/data/examples/ocr_example.html

File: doc/source/data/examples/ocr_example.ipynb [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a7854811",
+   "metadata": {},
+   "source": [
+    "# Extracting Information from Documents using Ray Datasets\n",
+    "\n",
+    "In this example, we will show you how to run optical character recognition (OCR) on a set of documents and analyze the resulting text with the natural language processing library spaCy. To make it more interesting, we will run the analysis on the [LightShot](https://www.kaggle.com/datasets/datasnaek/lightshot) dataset. It is a large publicly available OCR dataset with a wide variety of different documents, all of them screen shots of various forms. It is easy to replace that dataset with your own data and adapt the example to your own use cases!\n",
```
Comment (richardliaw) [inline doc/source/data/examples/ocr_example.ipynb:30]: > run optical character recognition (OCR) on a set of documents and analyze the resulting text with the natural language processing library spaCy. To make it more interesting, we will run the analysis on the dataset

I think this is missing a key component from the user perspective, which is to provide the answer to the question -- "Why do I need Ray for this example?"

**Review by clarkzinzow — COMMENTED** [review]
Nice example! OCR + NLP is a cool combo for an example.

I have a few high-level comments that I left on the appropriate bits of the PR, but to summarize:
1. We should try to use the Datasets I/O layer, if possible, instead of the out-of-band download + in-memory construction anti-pattern, even if it means that we need to host the data in a public S3 bucket ourselves.
2. The language detector inference should be vectorized as a `ds.map_batches()` call, since that will be more efficient on both the Datasets side and the spaCy side, and is Datasets' default recommendation for transformations.

File: doc/source/data/examples/ocr_example.ipynb [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+      "{'language': 'en', 'count()': 2}\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "\n"
+     ]
+    }
```
Comment (clarkzinzow) [inline doc/source/data/examples/ocr_example.ipynb:357]:
```suggestion
    }
```

File: doc/source/data/examples/ocr_example.ipynb [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+    "In this example, we will show you how to run optical character recognition (OCR) on a set of documents and analyze the resulting text with the natural language processing library spaCy. Running OCR on a large dataset is very computationally expensive, so using Ray for distributed processing can really speed up the processing. Ray Datasets makes it easy to compose the different steps of the pipeline, namely the OCR and the natural language processing.\n",
+    "\n",
+    "To make it more interesting, we will run the analysis on the [LightShot](https://www.kaggle.com/datasets/datasnaek/lightshot) dataset. It is a large publicly available OCR dataset with a wide variety of different documents, all of them screen shots of various forms. It is easy to replace that dataset with your own data and adapt the example to your own use cases!\n",
+    "\n",
+    "## Overview\n",
+    "\n",
+    "This tutorial will cover:\n",
+    " - Creating a Ray Dataset that represents the images in the dataset\n",
+    " - Running the computationally expensive OCR process on each image in the dataset in parallel\n",
+    " - Filtering out images that actually contain text\n",
```
[the remark this answers is not in the document]
Comment (pcmoritz) [reply doc/source/data/examples/ocr_example.ipynb]: Haha yes, English is hard! Should be fixed now :)

File: doc/source/_toc.yml [hunk 4]
```diff
@@ -51,6 +51,8 @@ parts:
                 title: Processing the NYC taxi dataset
               - file: data/examples/big_data_ingestion
                 title: Large-scale ML Ingest
+              - file: data/examples/ocr_example
+                title: Running OCR with Ray dataset
```
[the remark this answers is not in the document]
Comment (clarkzinzow) [reply doc/source/_toc.yml]: Nice, that's even better!

File: doc/source/data/examples/ocr_example.ipynb [hunk 5]
```diff
@@ -0,0 +1,477 @@
...
+    "           return LanguageDetector()\n",
+    "\n",
+    "        self.nlp.add_pipe('language_detector', last=True)\n",
+    "\n",
+    "    def __call__(self, row):\n",
+    "        row = dict(row)\n",
+    "        doc = self.nlp(row[\"text\"])\n",
+    "        row[\"language\"] = doc._.language[\"language\"]\n",
+    "        row[\"score\"] = doc._.language[\"score\"]\n",
+    "        return row\n",
```
[the remark this answers is not in the document]
Comment (pcmoritz) [reply doc/source/data/examples/ocr_example.ipynb]: Thanks, I batched it now!

File: doc/source/data/examples/ocr_example.ipynb [hunk 6]
```diff
@@ -0,0 +1,477 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "07ab8268",
+   "metadata": {},
+   "source": [
+    "### Running the OCR software on the data\n",
+    "\n",
+    "We first create a list `files` of absolute paths of the file names and then convert it into a Ray Dataset with the `ray.data.from_items` function. We can now run the `.map` function on this dataset of file names to run the actual OCR process on each file and convert the screen shots into text. Note that we only store the paths of the filenames and the OCR'ed text in the dataset to keep the size of the dataset manageble. If you have large binary blobs like images or videos, it can be beneficial to store them outside of the dataset and store only the extracted information in the dataset."
```
[the remark this answers is not in the document]
Comment (pcmoritz) [reply doc/source/data/examples/ocr_example.ipynb]: Implemented both suggestions!

File: doc/source/data/examples/ocr_example.ipynb [hunk 7]
```diff
@@ -0,0 +1,477 @@
...
+    "from glob import glob\n",
+    "import pytesseract\n",
+    "\n",
+    "files = glob(OCR_DATA + \"*\")\n",
+    "ds = ray.data.from_items(files)\n",
+    "\n",
+    "def perform_ocr(path):\n",
+    "    return {\"path\": path, \"text\": pytesseract.image_to_string(path)}\n",
+    "\n",
+    "results = ds.map(perform_ocr)"
```
[the remark this answers is not in the document]
Comment (pcmoritz) [reply doc/source/data/examples/ocr_example.ipynb:157]: fixed

File: doc/source/data/examples/ocr_example.ipynb [hunk 8]
```diff
@@ -0,0 +1,545 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "691077e0",
+   "metadata": {},
+   "source": [
+    "# Scaling OCR using Ray Datasets\n",
+    "\n",
+    "In this example, we will show you how to run optical character recognition (OCR) on a set of documents and analyze the resulting text with the natural language processing library [spaCy](https://spacy.io/). Running OCR on a large dataset is very computationally expensive, so using Ray for distributed processing can really speed up the processing. Ray Datasets makes it easy to compose the different steps of the pipeline, namely the OCR and the natural language processing. Ray Dataset's actor support also allows us to be more efficient by sharing the spaCy NLP context between several datapoints.\n",
```
Comment (clarkzinzow) [inline doc/source/data/examples/ocr_example.ipynb:30]:
```suggestion
    "In this example, we will show you how to run optical character recognition (OCR) on a set of documents and analyze the resulting text with the natural language processing library [spaCy](https://spacy.io/). Running OCR on a large dataset is very computationally expensive, so using Ray for distributed processing can really speed up the analysis. Ray Datasets makes it easy to compose the different steps of the pipeline, namely the OCR and the natural language processing. Ray Datasets' actor support also allows us to be more efficient by sharing the spaCy NLP context between several datapoints.\n",
```

File: doc/source/data/examples/ocr_example.ipynb [hunk 9]
```diff
@@ -0,0 +1,545 @@
...
+    "\n",
+    "To make it more interesting, we will run the analysis on the [LightShot](https://www.kaggle.com/datasets/datasnaek/lightshot) dataset. It is a large publicly available OCR dataset with a wide variety of different documents, all of them screen shots of various forms. It is easy to replace that dataset with your own data and adapt the example to your own use cases!\n",
```
Comment (clarkzinzow) [inline doc/source/data/examples/ocr_example.ipynb:32]:
```suggestion
    "To make it more interesting, we will run the analysis on the [LightShot](https://www.kaggle.com/datasets/datasnaek/lightshot) dataset. It is a large publicly available OCR dataset with a wide variety of different documents, all of them screenshots of various forms. It is easy to replace that dataset with your own data and adapt the example to your own use cases!\n",
```

File: doc/source/data/examples/ocr_example.ipynb [hunk 10]
```diff
@@ -0,0 +1,545 @@
...
+      "{'language': 'en', 'count()': 4}\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "\n"
+     ]
+    }
```
Comment (clarkzinzow) [inline doc/source/data/examples/ocr_example.ipynb:417]: This empty stderr block is still showing up
```suggestion
    }
```

File: doc/source/data/examples/ocr_example.ipynb [hunk 11]
```diff
@@ -0,0 +1,545 @@
...
+   "source": [
+    "results.take(10)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c6821410",
+   "metadata": {},
+   "source": [
+    "### Saving and loading the result of the OCR run\n",
```
Comment (clarkzinzow) [inline doc/source/data/examples/ocr_example.ipynb:199]: In this "save the intermediate results" section, we should note to the user that this isn't necessary, and that they could take the in-memory OCR results and apply the spaCy language detection without persisting to storage.

File: doc/source/data/examples/ocr_example.ipynb [hunk 12]
```diff
@@ -0,0 +1,545 @@
...
+    }
+   ],
+   "source": [
+    "from io import BytesIO\n",
+    "from PIL import Image\n",
+    "import pytesseract\n",
+    "\n",
+    "def perform_ocr(data):\n",
+    "    path, img = data\n",
+    "    return {\"path\": path, \"text\": pytesseract.image_to_string(Image.open(BytesIO(img)))}\n",
```
Comment (clarkzinzow) [inline doc/source/data/examples/ocr_example.ipynb:153]: To keep the wrapped line from making the indentation confusing, could we format this returned dictionary so it doesn't automatically wrap?

**Review by clarkzinzow — APPROVED** [review]
LGTM overall, just a few nits and suggestions. Thank you for making those previous changes, this example is awesome!

File: doc/source/data/examples/ocr_example.ipynb (hunk 11 above)
Comment (pcmoritz) [reply doc/source/data/examples/ocr_example.ipynb:199]: fixed

**Review by richardliaw — APPROVED** [review]

**Pull request merged**